flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,204 @@
+
+	private final PartitionTable inProgressPartitionTable = new PartitionTable();
+	private final PartitionTable finishedPartitionTable = new PartitionTable();
+
+	private Consumer<JobID> listener = jobId -> {};
+
+	/**	Tracks which jobs are still being monitored, to ensure cleanup in cases where tasks are finishing while
+	 * the jobmanager connection is being terminated. This is a concurrent map since it is modified by both the
+	 * Task (via {@link #notifyPartitionFinished(JobID, ResultPartitionID)}} and
+	 * TaskExecutor (via {@link #releaseAllFinishedPartitionsForJobAndMarkJobInactive(JobID)}) thread. */
+	private final Set<JobID> activeJobs = ConcurrentHashMap.newKeySet();
+
+	public JobAwareShuffleEnvironmentImpl(ShuffleEnvironment<?, G> backingShuffleEnvironment) {
+		this.backingShuffleEnvironment = Preconditions.checkNotNull(backingShuffleEnvironment);
+	}
+
+	@Override
+	public void setPartitionFailedOrFinishedListener(Consumer<JobID> listener) {
+		this.listener = listener;
+	}
+
+	@Override
+	public boolean hasPartitionsOccupyingLocalResources(JobID jobId) {
+		return inProgressPartitionTable.hasTrackedPartitions(jobId) || finishedPartitionTable.hasTrackedPartitions(jobId);
+	}
+
+	@Override
+	public void markJobActive(JobID jobId) {
+		activeJobs.add(jobId);
+	}
+
+	@Override
+	public void releaseFinishedPartitions(JobID jobId, Collection<ResultPartitionID> resultPartitionIds) {
+		finishedPartitionTable.stopTrackingPartitions(jobId, resultPartitionIds);
+		backingShuffleEnvironment.releasePartitions(resultPartitionIds);
+	}
+
+	@Override
+	public void releaseAllFinishedPartitionsForJobAndMarkJobInactive(JobID jobId) {
+		activeJobs.remove(jobId);
+		Collection<ResultPartitionID> finishedPartitionsForJob = finishedPartitionTable.stopTrackingPartitions(jobId);
+		backingShuffleEnvironment.releasePartitions(finishedPartitionsForJob);
+	}
+
+	/**
+	 * This method wraps partition writers for externally managed partitions and introduces callbacks into the lifecycle
+	 * methods of the {@link ResultPartitionWriter}.
+	 */
+	@Override
+	public Collection<ResultPartitionWriter> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup) {
+
+		final Collection<? extends ResultPartitionWriter> resultPartitionWriters = backingShuffleEnvironment
+			.createResultPartitionWriters(taskName, executionAttemptID, resultPartitionDeploymentDescriptors, outputGroup, buffersGroup);
+
+		// determine which partitions are externally managed, and whose writers need to be wrapped
+		final Set<ResultPartitionID> externallyManagedPartitions = resultPartitionDeploymentDescriptors.stream()
+			.filter(ResultPartitionDeploymentDescriptor::isManagedExternally)
+			// TODO: The descriptor should be extended to also contain the ResultPartitionID
```

Comment:
could extract the id from the ShuffleDescriptor in the RPDD.

Reply:
Yes, I also proposed this similar issue in #7835 , but it was not confirmed before, so it was closed then. I could submit this issue again.